src/rust/engine/rule_graph/src/builder.rs

```diff
@@ -272,6 +275,36 @@ impl<R: Rule> Builder<R> {
     self.finalize(pruned_edges_graph)
   }
 
+  ///
+  /// Validate that all rules have unique RuleIds.
+  ///
+  fn validate_rule_ids(&self) -> Result<(), String> {
+    let mut invalid_rule_ids: Vec<&RuleId> = self
+      .rules
+      .values()
+      .flatten()
+      .group_by(|rule| rule.id())
+      .into_iter()
+      .filter_map(|(rule_id, group)| {
+        if group.count() > 1 {
+          Some(rule_id)
+        } else {
+          None
+        }
+      })
```

Comment:
I note that https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.group_by only groups consecutive equal elements, i.e. `a, b, a` ends up with three groups of size 1.

Is `self.rules.values().flatten()` going to be sorted by `rule.id()`? If not, the docs of https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.duplicates suggest it might handle this well:


```suggestion
      .duplicates()
```

Reply:
Good call! They are not guaranteed to be sorted, and this would explain some weirdness.